nibabel/funcs.py

```diff
@@ -224,7 +224,18 @@ def as_closest_canonical(img, enforce_diag=False):
     # we need to transform the data
     arr = img.get_data()
     t_arr = apply_orientation(arr, ornt)
-    return img.__class__(t_arr, out_aff, img.header)
+    # Also apply the transform to the dim_info fields
+    new_hdr = img.header.copy()
+    new_dim = list(new_hdr.get_dim_info())
+    for idx, value in enumerate(new_dim):
```

Comment:
Sorry if I'm missing something obvious, by why would we not also permute the `None`s?

Maybe something like this:

```
    indices = ornt[:, 0].astype(np.intp)
    new_dim = np.array(new_hdr.get_dim_info(), dtype=object)[indices]
```